shell/server/drivers/external-ui-view.js

```diff
@@ -0,0 +1,319 @@
+// Sandstorm - Personal Cloud Sandbox
+// Copyright (c) 2014 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+var Future = Npm.require("fibers/future");
+var Promise = Npm.require("es6-promise").Promise;
+var Capnp = Npm.require("capnp");
+var Url = Npm.require("url");
+var Http = Npm.require("http");
+var Https = Npm.require("https");
+var ApiSession = Capnp.importSystem("sandstorm/api-session.capnp").ApiSession;
+
+var removeTrailingSlash = function (path) {
+  if (path.indexOf("/", path.length - 1) === -1) {
+    return path;
+  }
+
+  return path.slice(0, path.length - 1);
+};
+
+WrappedUiView = function (token, proxy, path) {
+  // TODO(someday): handle the fact that these proxies will be garbage collected every 2 minutes,
+  // even if it's in use.
+  this.token = token;
+  this.proxy = proxy;
+  this.path = path;
+};
+
+WrappedUiView.prototype.newSession = function (userInfo, context, sessionType, sessionParams, retryCount) {
+  if (sessionType !== ApiSession.typeId) {
+    throw new Error("SessionType must be ApiSession.");
+  }
+
+  retryCount = retryCount || 0;
+  var self = this;
+
+  return this.proxy.keepAlive().then(function () {
+    // ignore the passed userInfo and instead use the one associated with this token
+    // TODO(someday): handle permissions from the viewInfo
+    var session = self.proxy.uiView.newSession(
+        self.proxy.userInfo, context, sessionType, sessionParams).session.castAs(ApiSession);
+    return {session: new Capnp.Capability(new WrappedApiSession(session, self.path), ApiSession)};
+  }).catch(function (error) {
+    return self.proxy.maybeRetryAfterError(error, retryCount).then(function () {
+      return self.newSession(userInfo, context, sessionType, sessionParams, retryCount + 1);
+    });
+  });
+};
+
+WrappedApiSession = function (session, path) {
```

Comment:
Hmm it didn't occur to me that allowing a path in the API token implies that we need to wrap ApiSession. That's sad because we have to update this code any time ApiSession (or WebSession) gets a new method, and you're actually not even wrapping some of the existing methods now. Also sad for performance.

I don't think we have an actual use case for including a path in an API token, so how about we just disallow it and simplify this?